### ISSUE TITLE
npm based citation styles

### ISSUE THREAD

**Issue by lewisacidic** [citation-js/citation-js#6]
Hi Lars,

Really nice project! I had an idea about how to distribute CSL citation styles/locales in an NPM-friendly world, as very simple NPM packages.  I set up a [repo to produce these](https://github.com/richlewis42/citation-styles), and have uploaded a few useful ones:

- `style-nature`
- `style-apa`
- `style-mla`
- `style-chicago`
- `style-vancouver`
- `style-rsc`
- `locale-en-gb`
- `locale-en-us`
- `locale-es-es`
- `locale-de-de`
- `locale-fr-fr`

What do you think? This was just a trial to see if it works, but it seems to work nicely. Thought this project could use them too.

Rich

**Comment by larsgw** [comment]
That looks cool! It would definitely make version management easier, and make it clearer where some file size is coming from. Maybe it's a good idea to put it under a scope (like `@csl/style-nature` or `@citation-style-language/style-nature`). If so, we should probably discuss with the CSL people ([forum](https://discourse.citationstyles.org/)) first.

**Comment by larsgw** [comment]
I thought maybe someone took the long one just to be safe, maybe they registered `@csl/` in the same vein, I guess we'll see.

**Comment by richlewis42** [comment]
Yeah I guess it could well have been them - I assumed that csl as a 3 letter org was likely taken for another reason.

**Comment by rmzelle** [comment]
Hey, just happened to come across this. (CSL maintainer here)

I don't recall anybody closely affiliated with CSL claiming `@csl` on npm. And https://www.npmjs.com/~axept strongly suggests it's not CSL-related ("Client-Side Logger"):

>csl
>Description
>Client-Side Logger
>Publisher
>axept
>published 0.0.0-reserved • 2 years ago

I see the point of packaging up individual popular styles, although we would probably want a collection of all styles as well. I was just contemplating recently that CSL should implement some regular backup of our various repos outside of GitHub, and packaging could play a role in that. (@adam3smith)

**Comment by larsgw** [comment]
> I don't recall anybody closely affiliated with CSL claiming `@csl` on npm.

Someone could always ask support to see if the owner is willing to give it up, if you think that's useful. There are rules against squatting, but it may be a private organisation. [docs](https://www.npmjs.com/policies/disputes#organizations)

> And https://www.npmjs.com/~axept strongly suggests it's not CSL-related ("Client-Side Logger")

I don't really see the link between `~axept` and `@csl` other than the package, is there something I missed suggesting that? BTW, `~axept` does a *lot* of typo-squatting.

**Comment by rmzelle** [comment]
> I don't really see the link between `~axept` and `@csl` other than the package, is there something I missed suggesting that?

I thought we were talking about https://www.npmjs.com/package/csl. Is this about https://www.npmjs.com/org/csl?

**Comment by larsgw** [comment]
Yeah, I thought it'd make more sense to publish as `@csl/locale-..-..` and `@csl/style-...` but in `csl` would work too. The only slight problem is that some bundlers might pack the entire package, instead of just the styles and locales you want/need. However, getting the `csl` package should be a lot easier since it's definitely being squatted, and could be transferred immediately by npm support (see link above).

> The org looks more mysterious. No idea who registered that.

No, but it could be a private organisation and there are a lot of companies named "CSL" that could qualify for that.

**Comment by richlewis42** [comment]
Hi @rmzelle! Glad to see you found this - I was going to post this on the CSL forum once I've revisited the packaging a bit, but you're ahead of me!

My idea was to package the common styles separately as `.js` files (i.e. converted from the .csl file to JS objects, which `citeproc-js` can use natively), so any JS project that might need just one of the styles could easily install it and bundle it with their apps with native build tools. The use case that got me interested was for a blog powered by static site generators such as `gatsby`, where you would just install the style you wanted your references formatted with, configure appropriately, and the build tool would handle the rest. I'll let you know when I've got a working example!

You could certainly package all locales and styles in one big package, and users would be able to include only the files they need in any application bundle with a tree shaking build tool - however to install this package, users would have to download all the styles, which would end up pretty large.  Certainly a possibility though for people wanting very specific styles!

I've already asked npm to look into acquiring the `@csl` organisation.  They give the owner a month to get back, then they will release it. I think there's another week's wait, but haven't heard back negatively yet. We could also just name the individual packages `csl-style-*` and `csl-locale-*` if we don't get them, which should stop any collisions.

**Comment by richlewis42** [comment]
Just to update, I now have access to the `@csl` organisation!

**Comment by GedeonIsezerano** [comment]
Hi @lewisacidic @larsgw ,  I was wondering if the packages for citation styles/locales ever published I tried looking through npm using the @csl org with no luck.

**Comment by larsgw** [comment]
I haven't heard anything about it since but I am happy to help set it up if others want to work on it. I don't have access to the npm org.

**Comment by GedeonIsezerano** [comment]
I see. Thanks for letting me know.
I am new to the world of citations and styles but I would love to work on it if there are any guiding resources. I have been using the csl plugin with citation -js along with the csl files but the citation styles aren't quite right.

Here is a few styles I set up with the plugin and  the csl styles:

`import { Cite, plugins } from "citation-js";
import { chicagoStyle } from "../../citation_styles/chicago-author-date";
import { ieeeStyle } from "../../citation_styles/ieee-with-url";
import { mlaStyle } from "../../citation_styles/modern-language-association";

function addCSLTemplate(templateName: string, template: string) {
  let config = plugins.config.get("@csl");
  config.templates.add(templateName, template);
}

addCSLTemplate("Chicago 17th", chicagoStyle);
addCSLTemplate("IEEE", ieeeStyle);
addCSLTemplate("MLA 9th", mlaStyle);
`
Output (MLA 9th edition):

My output using the plugin: 
Smolyakov, A. I., Sabo, A., Yushmanov, P., & Putvinskii, S. (2021). On quasineutral plasma flow in the magnetic nozzle. Physics of Plasmas, 28(6). https://doi.org/10.1063/5.0050226

Expected output:
Smolyakov, A. I., et al. “On Quasineutral Plasma Flow in the Magnetic Nozzle.” Physics of Plasmas, vol. 28, no. 6, June 2021, p. 060701. DOI.org (Crossref), https://doi.org/10.1063/5.0050226.


Are there any other alternative solutions, or any ideas what could be causing the discrepancy in the outputs?

**Comment by larsgw** [comment]
How do you generate the output? It would be something like

```js
const cite = Cite(...)
cite.format('bibliography', { template: 'Chicago 17th' })
```

Otherwise it falls back to `apa`.

**Comment by larsgw** [comment]
That would make sense. I've opened #219 but I'm not sure when I am planning to make such breaking changes.
